### PR TITLE
refactor(freqai): cleanup some FreqAI models

### PR DESCRIPTION
**Pull request by jerome-benoit**
## Summary and changelog

* XGBoostRFClassifier, XGBoostClassifier: use test set weights at training (https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.XGBClassifier.fit)
* XGBoostRegressor: remove the train sets and weights from the eval sets and weights
* LightBGM*: align variables namespace, align `eval_weights` data structure type at variable affectation, no functional changes

### CODE REVIEW

**Comment by robcaulk** [comment]
This breaks existing user workflows, you've removed an entire evaluation metric from users' dashboards for no apparent reason. 

We are rejecting this.

**Comment by jerome-benoit** [comment]
> This breaks existing user workflows, you've removed an entire evaluation metric from users' dashboards for no apparent reason.

If it's a missing line in the summary explaining why adding the training sets to the evaluation sets while training a model is not the intended use of the scikit-learn compatible API and is theoretically not needed, it can be easily added, along with the theoretical justifications. 

Furthermore, I do not understand the rationale of a FOSS maintenance leadership to close a PR because a problematic small part of it which can lead to a potential breaking change at model evaluation, breaking changes are part of a lifecycle and can be postponed at PR review. The other changes are correct and will improve the training phase of two embedded classifiers.

**Comment by froggleston** [comment]
PRs can be reopened, so providing rationale as to why it should be, given the maintainer's feedback, is still an available route to you.

If you're aware that a PR introduces a breaking change, then that should be mitigated as much as possible ahead of time. I don't see any such mitigation in this PR so far.

**Comment by robcaulk** [comment]
> The other changes are correct and will improve the training phase of two embedded classifiers

Hello, I believe you have a core misunderstanding of how the XGBoost library is using the eval sets by default. I suppose the onus is now placed on me to teach you how the XGBoost library works, despite the open-source nature of the XGBoost algorithm and documentation.

Unless you add earlystopping to the `fit()` method, the eval sets are for logging only - hence why we have the training data eval as well for tensorboard logging.

It means that the changes you propose will not "improve the training phase" of these models at all, it will simply remove a log from their tensorboard for no apparent reason.

Lucky for you, as we discussed in the Freqtrade discord, the IFreqaiModel is analogous to the IStrategy. That means that you are not only able, but *encouraged* to change the eval and add `early_stopping_rounds`, or heck, why don't you add any other available parameter from the XGBoost library that you want to experiment with? Welcome to FreqAI.

**Comment by jerome-benoit** [comment]
> > The other changes are correct and will improve the training phase of two embedded classifiers
> 
> Hello, I believe you have a core misunderstanding of how the XGBoost library is using the eval sets by default. I suppose the onus is now placed on me to teach you how the XGBoost library works, despite the open-source nature of the XGBoost algorithm and documentation.

1) Aligning the default behavior of an API is the base of a sanely designed API. While also offering sane default code to copy. 
2) Aligning the default behavior of different software components targeted to the same use case is the base of a sane benchmark. And the base for actually building a meaningful evaluation dashboard across all the software components shipped by default in the API. In fact one would call it a fix. 
3) Removing the added evaluation sets to only one model code in the API shorten the training phase, which is called an improvement (while also addressing point 1) and 2)). 
4) The fact that the API is designed to be overloaded is not a valid rationale to not ship aligned behavior in an API. It's a best practice in software engineering. 

It would have been more respectful and productive and less condescending from a FOSS maintenership point of view to build a discussion around the PR changes than closing it.   

If the freqai people are fine by shipping an embedded API with unaligned behavior leading to benchmark bias in term of runtime measurement and training dashboard, and continuing to ship an API not following the most basic principles of API engineering, it's fine by me. I've already spent too much time in my life at dealing with uncooperative and stubborn FOSS maintenership.

File: docs/freqai-feature-engineering.md [hunk 1]
```diff
@@ -186,7 +186,7 @@ In total, the number of features the user of the presented example strategy has
 
 ### Gain finer control over `feature_engineering_*` functions with `metadata`
 
-All `feature_engineering_*` and `set_freqai_targets()` functions are passed a `metadata` dictionary which contains information about the `pair`, `tf` (timeframe), and `period` that FreqAI is automating for feature building. As such, a user can use `metadata` inside `feature_engineering_*` functions as criteria for blocking/reserving features for certain timeframes, periods, pairs etc.
+All `feature_engineering_*` and `set_freqai_targets()` functions are passed a `metadata` dictionary, which contains information about the `pair`, `tf` (timeframe), and `period` that FreqAI is automating for feature building. As such, a user can use `metadata` inside `feature_engineering_*` functions as criteria for blocking/reserving features for certain timeframes, periods, pairs etc.
```
Comment (robcaulk) [inline docs/freqai-feature-engineering.md:189]: Unrelated, remove from PR.

File: freqtrade/freqai/prediction_models/LightGBMClassifier.py [hunk 2]
```diff
@@ -28,20 +28,21 @@ def fit(self, data_dictionary: dict, dk: FreqaiDataKitchen, **kwargs) -> Any:
         :param dk: The datakitchen object for the current coin/model
         """
 
+        X = data_dictionary["train_features"].to_numpy()
```
Comment (robcaulk) [inline freqtrade/freqai/prediction_models/LightGBMClassifier.py:31]: Modifying variable names for unknown reasons does not constitute a change. Remove this file from PR.

File: freqtrade/freqai/prediction_models/LightGBMRegressor.py [hunk 3]
```diff
@@ -28,15 +28,16 @@ def fit(self, data_dictionary: dict, dk: FreqaiDataKitchen, **kwargs) -> Any:
         :param dk: The datakitchen object for the current coin/model
         """
 
+        X = data_dictionary["train_features"]
```
Comment (robcaulk) [inline freqtrade/freqai/prediction_models/LightGBMRegressor.py:31]: Modifying variable names for unknown reasons does not constitute a change. Remove this file from PR.

File: freqtrade/freqai/prediction_models/XGBoostRegressor.py [hunk 4]
```diff
@@ -36,8 +36,8 @@ def fit(self, data_dictionary: dict, dk: FreqaiDataKitchen, **kwargs) -> Any:
             eval_set = None
             eval_weights = None
         else:
-            eval_set = [(data_dictionary["test_features"], data_dictionary["test_labels"]), (X, y)]
-            eval_weights = [data_dictionary["test_weights"], data_dictionary["train_weights"]]
+            eval_set = [(data_dictionary["test_features"], data_dictionary["test_labels"])]
```
Comment (robcaulk) [inline freqtrade/freqai/prediction_models/XGBoostRegressor.py:39]: Breaking other people's logs without any rationale is not ok, remove this file from PR.

**Comment by robcaulk** [comment]
Let's just keep in mind what is going on here. You are demanding to make breaking changes that won't actually change any of the performance of the models. Meanwhile, in some of your files, you've only modified variable names, and they aren't even consistent across the board.

None of your changes have any effect on any models. The only change you've made that has any material effect is the one that breaks existing users tensorboard logs.

We've pointed all of this out to you and you are doubling down, telling us in the Freqtrade discord that we "have no idea what we are doing".

If you want this PR to go through, remove all unnecessary changes (which appears to mean that you need to remove the majority of files you've changed), and revert your change that breaks people's tensorboard logs.

> It would have been more respectful

I'm sorry, I've observed that **you have a shocking pattern of treating a range of volunteers in multiple discord servers with near total disrespect**. The audacity to demand respect now, on a PR that breaks people's logs - is unbelievable to me.

Here is some public context regarding "respect". @jerome-benoit joined the FreqAI discord and:

- 1st hour: Jerome demanded that other users make github repos of their personal strategies "because that is the proper way to iterate code". This is him talking to volunteers who are already generously sharing their files publicly (cough cough).
- 2nd hour: Jerome says freqai has bad "errors" and "there are bugs". Volunteers requested errors so that they could help him with his problem. but didnt give any errors and ignored them.
- 3rd hour: Jerome made a feature request about adding normalization as a feature. In response, volunteers politely gave him help by pointing him to the normalization feature that already exists, and happens to be the first line in the docs. Volunteers included exact screenshots to help him with his request. Instead of any sign of appreciation to the volunteers, he immediately started criticizing docs. Let's keep in mind that these docs were written by volunteers, aimed at helping people just like our friend @jerome-benoit.

- Jerome wrapped up his tour of volunteer disrespect by demanding that the volunteers make a PR for him to "fix a link" that is not broken.

He is treating all of us volunteers like his own ChatGPT.
